`HUB` consistency arguments

OlivierBBB opened the pull request.

OlivierBBB committed:
release: bump version to beta-v1.2.0-rc2 and update changelog

OlivierBBB committed:
Merge branch 'arith-dev' into rel/beta-v1.2.0-rc2

OlivierBBB committed:
feat: disable rust corset validator + changed constraints commit

OlivierBBB committed:
feat: turn on HUB consistency constraints

OlivierBBB committed:
fix: ras (go-corset flags as suggested by @David.Pearce)

OlivierBBB committed:
fix: Besu reverets accrued states gefore it traceEndTransaction

This fix enforces that the sender be warm in all cases during TX_FINL

OlivierBBB committed:
fix: the CFI must always be printed (and defults to 0)

Previously we were writing

  account/CODE_FRAGMENT_INDEX

only if requiresRomLex. This is wrong and breaks account consistency
constraints.

OlivierBBB committed:
ras: dead code

OlivierBBB committed:
fix: CallFrame now initialized with outputDataRange pointing to them

OlivierBBB committed:
fix: pre-wiping snapshot now uses canonical()

fixes issue where SELFDESTRUCT's would have old and new
AccountSnapshot's be empty on the account wiping account-row.

OlivierBBB committed:
fix:ras

OlivierBBB committed:
fix: constraints commit update

fixes various acp_ related bugs

OlivierBBB committed:
fix: mix-up in swapStackPattern between "top" and "bottom" values

OlivierBBB committed:
ras: banal tests + spotless

OlivierBBB committed:
fix: subtle point about scn/CALL_SMC_FAILURE_WILL_REVERT for self calls

OlivierBBB committed:
ras

OlivierBBB commented:
https://github.com/Consensys/linea-tracer/pull/1746/commits/26c391210aacaaf90ed9a991df8e5fc9d472ff5c addresses https://github.com/Consensys/linea-specification/issues/83#issue-2816073556.

OlivierBBB committed:
fix: no more squashing return data

The point being: our MemoryRange object now should always remember the
relevant context number. MemoryRange.EMPTY, however, doesn't remember
it. This is problematic when encountering a context row after the
context row which updated the return data after an exception.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -733,8 +733,6 @@ public void tracePostExecution(MessageFrame frame, Operation.OperationResult ope
           .exceptionalContextFragment(ContextFragment.executionProvidesEmptyReturnData(this));
-      this.squashCurrentFrameOutputData();
-      this.squashParentFrameReturnData();
     }
```
`squashParentFrameReturnData` replaces the value of `MemorySpan returnData` provided to the parent frame in `exceptionalContextFragment`, which is
```java
MemorySpan(currentFrame.contextnumber())
```
which has a nonzero context number, with
```java
MemorySpan.EMPTY
```
which has a context number of `0`. This was a mistake dating back to the time when we had both a `returnDataContextNumber` and a `returnDataSpan` rather than the current system of having both united in a `MemorySpan` object.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -1100,7 +1098,7 @@ public void squashCurrentFrameOutputData() {
 
   public void squashParentFrameReturnData() {
-    callStack.parentCallFrame().outputDataRange(MemoryRange.EMPTY);
+    callStack.parentCallFrame().returnDataRange(MemoryRange.EMPTY);
```
Typo

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/account/AccountFragment.java:
```diff
@@ -186,10 +186,12 @@ public void resolveAtEndTransaction(
+              newState.address(), newState.deploymentNumber(), newState.deploymentStatus());
+    } catch (RuntimeException e) {
+      codeFragmentIndex = 0;
+    }
```
CFI's are required on **all** account-rows, but they are **looked up** only on those where `requiresRomlex = true`.

letypequividelespoubelles replied:
I still don't understand the need for this change : in case `requiresRomlex` is false, the previous method gives 0 and shouldn't raise an exception

OlivierBBB replied:
> in case requiresRomlex is false, the previous method gives 0 and shouldn't raise an exception

I may have mixed things up then. I vaguely remember getting an exception at some point, but I may be misremembering.

OlivierBBB replied:
The reason, as discussed earlier, seems to be that when we request the CFI of an account with either empty code or whose byte code does not get loaded into the ROM it may throw an exception, which is what I remember. I may change it to
```java
if (cfiRequired) {
  // old way
} else {
  // above way with error handling
}
```

to make sure we don't have "required CFI requests" that don't find a match in the ROM_LEX map.

OlivierBBB replied:
We now check that if `cfiRequired = true` that getting the CFI doesn't throw a run time exception.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxFinalizationSection.java:
```diff
@@ -54,6 +54,12 @@ public void resolveAtEndTransaction(
+    // TX_FINL");
+    // checkArgument(
+    //     hub.messageFrame().isAddressWarm(hub.txStack().current().getEffectiveRecipient()),
+    //     "The recipient should be warm at TX_FINL");
```
This doesn't work. In Besu you
```java
MessageFrame.revert()
```
before you `traceEndTransaction`. As a consequence the sender etc ... aren't warm at transaction end if either there was an `EXCEPTIONAL_HALTING_CONDITION` of any kind or if the final opCode of the root frame was a `REVERT`.

letypequividelespoubelles replied:
better delete the commented code then ?

OlivierBBB replied:
Done

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxFinalizationSection.java:
```diff
@@ -123,7 +129,7 @@ private void setSnapshots(Hub hub, WorldView world) {
         senderIsCoinbase(hub)
             ? coinbaseGasRefund.deepCopy()
-            : AccountSnapshot.canonical(hub, world, senderAddress);
+            : AccountSnapshot.canonical(hub, world, senderAddress).turnOnWarmth();
```
The comment on coinbase address warmth at transaction end, in particular in case of address collisions, explains why we do this.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:
```diff
@@ -605,8 +605,14 @@ private void completeEoaSuccessWillRevert(Hub hub) {
+      calleeThird = calleeSecondNew.deepCopy().setDeploymentNumber(hub);
+      calleeThirdNew = calleeFirst.deepCopy().setDeploymentNumber(hub);
+    }
+
```
See https://github.com/Consensys/linea-specification/issues/83 for an explanation.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/callstack/CallFrame.java:
```diff
@@ -203,6 +203,7 @@ public void rememberGasNextBeforePausing(Hub hub) {
     this.parentId = parentId;
     this.callDataRange = callDataRange;
     this.returnAtRange = returnAtRange;
+    this.outputDataRange = new MemoryRange(contextNumber);
```
The `outputDataRange`, if not overwritten nontrivially at `traceContextExit`, should at least remember its own context number, as that will always be correct and provides its parent frame at least with the correct `returnDataContextNumber` by means of `returnDataRange.contextNumber()`.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/callstack/CallFrame.java:
```diff
@@ -312,9 +313,4 @@ public OpCode getOpCode() {
-  /** Get to fucking work GitHub */
-  public void squashReturnData() {
-    returnDataRange(MemoryRange.EMPTY);
-  }
```
A relic of past frustration when GitHub runners were down for a few days 😅

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/stack/Stack.java:
```diff
@@ -153,17 +153,18 @@ private void dup(MessageFrame frame, StackContext pending) {
-    Bytes val1 = getStack(frame, 0);
-    Bytes val2 = getStack(frame, depth);
+    Bytes topValue = getStack(frame, 0);
+    Bytes botValue = getStack(frame, depth);
```
Simple mix up. Was invisible before stack consistency constraints.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/TransactionProcessingMetadata.java:
```diff
@@ -180,9 +180,7 @@ public void completeLineaTransaction(
-      destructedAccountsSnapshot.add(
-          AccountSnapshot.fromAddress(
-              address, true, hub.deploymentNumberOf(address), hub.deploymentStatusOf(address)));
+      destructedAccountsSnapshot.add(AccountSnapshot.canonical(hub, address));
```
Using `canonical` gives us access to the correct code, nonce, etc ...

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/corset/GoCorsetValidator.java:
```diff
@@ -95,7 +95,8 @@ private void configGoCorset() {
     ArrayList<String> options = new ArrayList<>();
     // Determine options to use (either default or override)
-    String flags = System.getenv().getOrDefault("GOCORSET_FLAGS", "-w --air");
+    String flags =
```
What's the diff ?

OlivierBBB replied:
The flags were suggested by @DavePearce

letypequividelespoubelles reviewed (approved):
LGTM except the change about the CFI I don't understand

DavePearce committed:
fix github action workflows

Since Rust corset cannot be used to check the constraints any longer
(i.e. due to unsupported features being used), it is no longer used in
the `linea-constraints` Makefile.  Thus, we no longer separate the
`zkevm.bin` from the `zkevm.go.bin` (and likewise for the reference
tests).

As a result of all this, the github actions workflow needed to be
updated to remove attempts to run the original corset tool, and also to
default back to using the `zkevm.bin` instead of `zkevm.go.bin` for
`go-corset`.

DavePearce committed:
tweak naming of github action workflows

This just tweaks the naming of workflows to remove explicit references
to go-corset.  Specifically, no need any more to distinguish
"unit-tests" from "go-corset-unit-tests", etc.

OlivierBBB committed:
fix: removed obsolete (commented) checks

DavePearce committed:
update `go-corset` github action to latest

This updates the github action to always use the latest version of
`go-corset`.  Since we're in the process of debugging the new defcompute
feature, this seems reasonable.

OlivierBBB committed:
ras: renaming PostTransactionDefer => EndTransactionDefer

OlivierBBB committed:
feat: new Defer type: AfterTransactionFinalizationDefer

Used by SELFDESTRUCT's

OlivierBBB committed:
wip: have SELFDESTRUCT use AfterTransactionFinalizationDefer

OlivierBBB committed:
ras

OlivierBBB committed:
fix: SELFDESTRUCT, wipe() and account wiping Account Rows

OlivierBBB committed:
fix: clearing of the "after TX_FINL defer" registry

Thanks @François : )

OlivierBBB committed:
fix: fixing CALLs (SMC_FAILURE_WONT_REVERT case)

The deployment status gets reverted, the deployment number is
monotonous.

OlivierBBB committed:
fix: typo

OlivierBBB committed:
fix: use the most up to date WorldUpdater to get balances for self destructs

This solved some tests within _601 and _612 where the coinbase address
would self destruct and display zero value before that event, rather
than the gas fee reward.

OlivierBBB committed:
ras: use intended values in GasLimitTest

OlivierBBB committed:
fix: remember init code of reverted deployment transactions

OlivierBBB committed:
fix: ignore tests involving block_reward

We ingore all tests where the payment of the (London) 2 Eth block reward
breaks account consistency constraints.

OlivierBBB committed:
ras

OlivierBBB committed:
fix: get the right warmth of the COINBASE address at TX_FINL

OlivierBBB committed:
ras

OlivierBBB committed:
fix: for CALL_EOA_SUCCESS_WILL_REVERT CallSection

This fixes the case of CALLs of the kind

  scenario/CALL_EOA_SUCCESS_WILL_REVERT

The above was failing when the caller was undergoing deployment. Using
the `setDeploymentInfo(Hub)` method led to using the updated deployment
status (false), and hence seeing the init code as deployed code. This
broke in particular the constraints involving the CODEHASH.

OlivierBBB committed:
fix: not sure what that change represents nor where it comes from

OlivierBBB committed:
fix: constraints update

OlivierBBB committed:
ras: constraints commit update

OlivierBBB committed:
ras: spotless + a tiny fix

OlivierBBB committed:
fix: constraints commit update

lorenzogentile404 commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/SelfdestructSection.java:
```diff
@@ -268,18 +263,14 @@ public void resolveAtEndTransaction(
+      selfdestructScenarioFragment.setScenario(SELFDESTRUCT_WONT_REVERT_NOT_YET_MARKED);
 
+      accountWipingNew = accountWiping.deepCopy();
       // the hub's defers.resolvePostTransaction() gets called after the
```
Is `accountWipingNew` supposed to be an exact copy of `accountWiping` when used to initialize:

```
   final AccountFragment accountWipingFragment =
          hub.factories()
              .accountFragment()
              .make(
                  accountWiping,
                  accountWipingNew,
                  DomSubStampsSubFragment.selfdestructDomSubStamps(hub, hubStamp));
```
?

OlivierBBB replied:
I don't remember the exact reason for doing things this way here, but if you look below we
```java
hub.defers().scheduleForAfterTransactionFinalization(this);
```
and at `traceEndTransaction`, after resolving everything in `endTransactionDefer` we will
```java
resolveAfterTransactionFinalization(...)
```
at which point we `accountWipingNew.wipe()`. I guess when doing the `deepCopy` we may get some values from `accountWiping` that won't get wiped.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:
```diff
@@ -199,6 +199,8 @@ public CallSection(Hub hub, MessageFrame frame) {
     callerFirst = canonical(hub, callerAddress);
     calleeFirst = canonical(hub, calleeAddress);
 
+    Address failureAddress = Address.fromHexString("0xc07f1349a887643be65b34e234e1b3161f62dc30");
```
?

letypequividelespoubelles replied:
I guess it was debugging code ?

letypequividelespoubelles replied:
@OlivierBBB

OlivierBBB replied:
Yes, fixing this now.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/AccountSnapshot.java:
```diff
@@ -156,17 +156,10 @@ public AccountSnapshot deepCopy() {
-        Bytecode.EMPTY,
-        deploymentInfo.deploymentNumber(address),
-        deploymentStatus);
+    this.nonce(0).balance(Wei.ZERO).code(Bytecode.EMPTY).setDeploymentInfo(deploymentInfo);
```
The `wipe` method must be invoked after the `SELFDESTRUCT`'ing account's deployment information underwent a `freshDeployment`.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -534,9 +534,9 @@ public void traceEndTransaction(
-
+    txStack.current().completeLineaTransaction(this, world, isSuccessful, logs, selfDestructs);
     defers.resolveAtEndTransaction(this, world, tx, isSuccessful);
+    defers.resolveAfterTransactionFinalization(this, world);
```
The only items that are in the `resolveAfterTransactionFinalization` defer registry are the "account wiping rows" of successful `SELFDESTRUCT`'s.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -745,7 +743,10 @@ public void tracePostExecution(MessageFrame frame, Operation.OperationResult ope
+          isExceptional() || opCode() == REVERT
+              ? txStack.current().coinbaseWarmthAfterTxInit(this)
+              : frame.isAddressWarm(coinbaseAddress);
     }
```
If the root context reverts Besu's frame will `revert` and reset account warmths. This will clash with conventions where e.g. the sender is warm at transaction end in the ZKEVM, regardless of circumstances. This isn't an issue for the sender but it may be an issue for the coinbase in case the coinbase is either the sender or the recipient.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/defer/DeferRegistry.java:
```diff
@@ -119,16 +129,23 @@ public void scheduleForPostRollback(PostRollbackDefer defer, CallFrame callFrame
     }
-    postTransactionDefers.clear();
+    endTransactionDefers.clear();
+  }
```
Nothing happens here, it's just the renaming of `postTransactionDefers` to `endTransactionDefers`.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/defer/DeferRegistry.java:
```diff
@@ -119,16 +129,23 @@ public void scheduleForPostRollback(PostRollbackDefer defer, CallFrame callFrame
+    for (AfterTransactionFinalizationDefer defer : afterTransactionFinalizationDefers) {
+      defer.resolveAfterTransactionFinalization(hub, worldView);
+    }
+    afterTransactionFinalizationDefers.clear();
```
Wiping this registry is important 🙂.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/account/AccountFragment.java:
```diff
@@ -154,8 +154,8 @@ public Trace trace(Trace trace) {
         .pAccountDeploymentStatus(oldState.deploymentStatus())
+        .pAccountDeploymentNumberNew(newState.deploymentNumber())
         .pAccountDeploymentStatusNew(newState.deploymentStatus())
         .pAccountDeploymentNumberInfty(deploymentNumberInfinity)
```
Just reordering.

OlivierBBB committed:
ras: renaming

OlivierBBB committed:
fix: added check wrt getting CFI when requiresRomLex = true

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -202,6 +202,12 @@ public void resolveAtEndTransaction(
+            .deploymentStatus(true)
+            .code(recipientValueReceptionNew.code());
+      }
+
```
This is required deployment transactions get converted to either
- deployed byte code
- no deployed byte code

and this requires a transition from the state of the account where there is initialization code (`recipientValueReceptionNew.code()`) to one where there isn't. We thus grab the initialization code to fill the "pre-state" with it.

OlivierBBB committed:
ras: removing redundant debugging code

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxSkipSection.java:
```diff
@@ -112,7 +112,7 @@ public void resolveAtEndTransaction(
     senderNew = canonical(hub, world, sender.address(), isPrecompile(sender.address()));
     recipientNew = canonical(hub, world, recipient.address(), isPrecompile(recipient.address()));
-    coinbaseNew = canonical(hub, world, coinbase.address(), isPrecompile(recipient.address()));
+    coinbaseNew = canonical(hub, world, coinbase.address(), isPrecompile(coinbase.address()));
```
typo

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:
```diff
@@ -575,11 +575,11 @@ private void completeAbortWillRevert(Hub hub) {
-    calleeSecond = reEntryCalleeSnapshot.deepCopy().setDeploymentInfo(hub);
-    calleeSecondNew = calleeFirst.deepCopy().setDeploymentInfo(hub);
+    calleeSecond = reEntryCalleeSnapshot.deepCopy().setDeploymentNumber(hub);
+    calleeSecondNew = calleeFirst.deepCopy().setDeploymentNumber(hub);
```
The deployment number may have evolved, but it is the deployment status we are walking back.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/call/CallSection.java:
```diff
@@ -630,11 +636,11 @@ private void completeSmcOrPrcSuccessWillRevert(Hub hub) {
-    calleeSecond = calleeFirstNew.deepCopy().setDeploymentInfo(hub);
-    calleeSecondNew = calleeFirst.deepCopy().setDeploymentInfo(hub);
+    calleeSecond = calleeFirstNew.deepCopy().setDeploymentNumber(hub);
+    calleeSecondNew = calleeFirst.deepCopy().setDeploymentNumber(hub);
```
The deployment number may have evolved, but it is the deployment status we are walking back.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/SelfdestructSection.java:
```diff
@@ -293,13 +284,24 @@ public void resolveAtEndTransaction(
+        .deploymentUpdateForSuccessfulSelfDestruct(selfdestructor.address());
 
-    this.addFragment(finalUnexceptionalContextFragment);
+    accountWipingNew.wipe(hub.transients().conflation().deploymentInfo());
```
This is the raison d'être of the `AfterTransactionFinalization` defer registry. It is so that we don't raise the deployment number of `SELFDESTRUCT`'ing accounts prior to executing the `TX_FINL` phase. This messes things up in case the `COINBASE` is among the `SELFDESTRUCT`'ing accounts.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/tables/instructionDecoder/InstructionDecoder.java:
```diff
@@ -126,7 +125,7 @@ public void commit(List<MappedByteBuffer> buffers) {
       traceBillingSettings(op, trace);
       trace
-          .opcode(Bytes.ofUnsignedInt(i))
+          .opcode(UnsignedByte.of(i))
```
@DavePearce I've had to correct this a couple of times here and there. Just letting you know, not sure if it's a corset change that's at the source of this change in signature.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/TransactionProcessingMetadata.java:
```diff
@@ -175,14 +176,16 @@ public void setPreFinalisationValues(
-      destructedAccountsSnapshot.add(
-          AccountSnapshot.fromAddress(
-              address, true, hub.deploymentNumberOf(address), hub.deploymentStatusOf(address)));
+      destructedAccountsSnapshot.add(AccountSnapshot.canonical(hub, world, address));
```
This was a very stupid oversight / temporary solution to a problem that only became noticeable with account consistency on. The `fromAddress` assigns empty code, zero balance, zero nonce etc. to the `AccountSnapshot` it is creating. Unsurprisingly this broke account consistency constraints.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/types/TransactionProcessingMetadata.java:
```diff
@@ -328,4 +331,23 @@ public void captureUpdatedInitialRecipientAddressDeploymentInfoAtTransactionStar
+        || coinbaseIsPrecompile
+        || coinbaseIsSender
+        || coinbaseIsRecipient;
+  }
```
This method logs whether the `COINBASE` address will be warm at the end of `TX_INIT`. This is required in case of a `REVERT`'ing transaction with an address collision. It will be made obsolete by [EIP-3651](https://eips.ethereum.org/EIPS/eip-3651).

OlivierBBB committed:
ras: documentation

OlivierBBB committed:
spotless

lorenzogentile404 reviewed (changes requested):
`accountWipingNew.wipe(hub.transients().conflation().deploymentInfo());` returns a new `AccountSnapshot` that is not used.

OlivierBBB committed:
merge arith-dev into HUB-consistency-arguments

OlivierBBB commented:
> `accountWipingNew.wipe(hub.transients().conflation().deploymentInfo());` returns a new `AccountSnapshot` that is not used.

As discussed on slack, this is not the case.

DavePearce committed:
remove `corset-unit-tests`
